modules/999-helm/hooks/metrics.go

```diff
@@ -0,0 +1,407 @@
+
+	total, err := processHelmReleases(k8sCurrentVersion, input, input.Snapshots["helm3_releases"])
+	if err != nil {
+		return err
+	}
+	input.MetricsCollector.Set("helm_releases_count", float64(total), map[string]string{"helm_version": "3"})
+
+	total, err = processHelmReleases(k8sCurrentVersion, input, input.Snapshots["helm2_releases"])
+	if err != nil {
+		return err
+	}
+	input.MetricsCollector.Set("helm_releases_count", float64(total), map[string]string{"helm_version": "2"})
+
+	return nil
+}
+
+func processHelmReleases(k8sCurrentVersion *semver.Version, input *go_hook.HookInput, snapshots []go_hook.FilterResult) (uint32, error) {
+	var totalDeployedReleases uint32
+	// get helm manifests
+	for _, sn := range snapshots {
+		if sn == nil {
+			continue
+		}
+
+		helmRelease := sn.(*helmRelease)
+
+		if helmRelease.StatusDeployed {
+			totalDeployedReleases++
+		}
+
+		d := yaml.NewDecoder(strings.NewReader(helmRelease.Release.Manifest))
+
+		for {
+			resource := new(manifest)
+			err := d.Decode(&resource)
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					break
+				}
+
+				return 0, err
+			}
+
+			if resource == nil {
+				continue
+			}
+
+			incompatibility, k8sCompatibilityVersion := storage.CalculateCompatibility(k8sCurrentVersion, resource.APIVersion, resource.Kind)
+			if incompatibility > 0 {
+				input.MetricsCollector.Set("resource_versions_compatibility", float64(incompatibility), map[string]string{
+					"helm_release_name":      helmRelease.Release.Name,
+					"helm_release_namespace": helmRelease.Release.Namespace,
+					"k8s_version":            k8sCompatibilityVersion,
+
+					"resource_name":      resource.Metadata.Name,
+					"resource_namespace": resource.Metadata.Namespace,
+					"kind":               resource.Kind,
+					"api_version":        resource.APIVersion,
+				}, metrics.WithGroup("helm_deprecated_apiversions"))
+			}
+
+		}
+	}
+
+	return totalDeployedReleases, nil
+}
+
+type helmRelease struct {
+	StatusDeployed bool
+	Release        *release
+}
+
+// protobuf for handling helm2 releases - https://github.com/helm/helm/blob/47f0b88409e71fd9ca272abc7cd762a56a1c613e/pkg/proto/hapi/release/release.pb.go#L24
+type release struct {
+	Name      string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name,proto3"`
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,8,opt,name=namespace,proto3"`
+	Manifest  string `json:"manifest,omitempty" protobuf:"bytes,5,opt,name=manifest,proto3" `
+}
+
+type manifest struct {
+	APIVersion string `json:"apiVersion" yaml:"apiVersion"`
+	Kind       string `json:"kind" yaml:"kind"`
+	Metadata   struct {
+		Name      string `json:"name" yaml:"name"`
+		Namespace string `json:"namespace" yaml:"namespace"`
+	} `json:"metadata" yaml:"metadata"`
+}
+
+// k8s version: APIVersion: Kind
+type unsupportedVersionsStore map[string]unsupportedAPIVersions
+
+func (uvs unsupportedVersionsStore) getByK8sVersion(version *semver.Version) (unsupportedAPIVersions, bool) {
```

Comment:
Because this method is used inside the `CalculateCompatibility`. And `CalculateCompatibility` is the entrance method of store object to make a calculations which is used inside hook handler.
Sure we don't need it as exported function but I separate User interface of structs in a such way